### ISSUE TITLE
support for cross-platform node_modules aliases and extensions 

### ISSUE THREAD

**Issue by alidlo** [thheller/shadow-cljs#650]
So I was using `:react-native` build with expo  to start both my native and web app. I didn't consider whether I should be using the `:browser`  build until now, in order to use `:modules`  build option.

The issue i'm running into is that now that I changed to the `:browser` build, shadow is loading `react-native` dependencies that before were being handled by expo's Webpack config (i.e. `react-native` was being aliased as `react-native-web` and `.web.js` files were being prioritized over `.js` ones. )

A lot of react native packages are adding web support via `.web` extension, but they expect the bundler to pick up the `.web.js` file over regular `.js` ones.

not sure if the best option is to add a `web-modules` section to the `:react-native` build (since I guess shadow doesn't load dependencies in this situation, it lets js bundler handle it) or to add support for prioritizing different platform extensions depending on the build

**Comment by thheller** [comment]
Sorry, but I don't quite follow what you are asking here.

You seem to be talking about 2 different topics and I can't quite put together what the actual question is. What do `:modules` have to do with `.web.js`?

Would help to have an actual example of what you have now and what you would like to have instead with some actual code?

**Comment by alidlo** [comment]
So I was stating that my app was working fine because I was using the `:react-native` build for web (letting expo's webpack configuration handle bundling). But, because I'd like to code split my cljs web code, I realized I needed to switch to the `:browser` build and let shadow-cljs handle bundling.

The issue is that shadow-cljs does not have the same support for cross-platform js code conventions that react-native packages rely on.

It's actually two issues, which is perhaps why I wasn't fully clear above.

The first one is that the way [react-native-web](https://github.com/necolas/react-native-web) works is it's a parallel implementation of react-native for web. So one's react-native code can work in web as long as they alias `react-native` imports to `react-native-web`. Please correct me if I'm wrong, but the `:build-options` alias is only intended for cljs code, not for js code, and so not sure how to support this behavior for a cljs web app.

The second issue is that, in cases where a react native package needs different implementations per platform, it relies on platform specific extensions being picked up. for example, [react-native-gesture-handler](https://github.com/software-mansion/react-native-gesture-handler) by default has `.js` for native and web code, and `.web.js` for web-only code. So when a `.web.js` file is provided, it expects you to configure your bundler so that the `.web.js` file is picked up before the `.js` one. 

So, to recap, two issues: 1) supporting aliases for javascript code 2) supporting platform extensions resolutions for javascript code. 

Are you open to exposing ability to configure these two compiler options for javascript code? Or, since shadow-cljs already supports react-native, at least supporting the cross-platform bundler conventions already used by react-native community.

Let me know if there's another approach you'd recommend to get around this. I'll try and add a minimal reproduction later, but wanted to provide more context in this issue meanwhile.

**Comment by thheller** [comment]
Ok, now I understand a bit better.

However I have very strong doubts that this is ever gonna work in any reasonable manner. I'm not sure about the overall state of react-native-web packages but react-native packages are pretty much built assuming you are going to be using metro. So they don't even put standard JS code into their packages and instead use whatever metro supports (like flow). I very much assume that the react-native-web story is just going to assume webpack.

shadow-cljs already supports configuring the extensions so you can configure `:js-options {:extensions [".web.js" ".js" ".json"]}` in your build config and it should start picking up `.web.js` files if they exist.

I doubt that'll get you very far though. It is very likely the packages are also gonna assume webpack processing CSS and so on. I wouldn't be surprised if you run into `.ts` files or flow JS.

**Comment by alidlo** [comment]
yeah, fortunately, there's new tools like https://github.com/react-native-community/bob to encourage devs to build their code for different platforms (you can have `main`, `module`, and `react-native` targets in your `package.json`). 

Thanks for letting me know about the js extension option, will try that out.

**Comment by thheller** [comment]
You may also configure `:js-options {:entry-keys ["react-native" "module" "browser" "main"]}`. The current default is `["browser" "main"]` though. There are often issues when mixing CommonJS<->ESM code so `module` isn't used by default.

**Comment by alidlo** [comment]
I was able to get my react-native-web app working in `:browser` build with these js options:
```
:js-options {:extensions [".web.js" ".js" ".json"]
                :resolve    {"react-native" {:target :npm
                                             :require "react-native-web"}}}
```

thanks for the help @thheller
